File: resources/android/StepperRenderer.kt
package dev.firstlightui.plugins.firstlight_ui.ui

import androidx.compose.foundation.isSystemInDarkTheme
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import com.nativephp.mobile.ui.nativerender.NativeUIBridge
import com.nativephp.mobile.ui.nativerender.NativeUINode
import com.nativephp.mobile.ui.nativerender.NativeUITree
import com.nativephp.plugins.native_ui.NativeUITheme

data class StepperRendererEvent(val callbackId: Int, val nodeId: Int) {
    val wireName = "PRESS"
}

class StepperRendererEvents(private val send: (StepperRendererEvent) -> Unit) {
    fun dispatch(event: StepperRendererEvent) = send(event)

    companion object {
        val native = StepperRendererEvents { event ->
            NativeUIBridge.sendPressEvent(event.callbackId, event.nodeId)
        }
    }
}

data class StepperRendererConfiguration(
    val nodeId: Int,
    val displayValue: String,
    val label: String,
    val helper: String,
    val error: String,
    val disabled: Boolean,
    val canDecrement: Boolean,
    val canIncrement: Boolean,
    val decrementCallback: Int,
    val incrementCallback: Int,
    val accessibilityLabel: String,
    val accessibilityHint: String,
) {
    constructor(node: NativeUINode) : this(
        nodeId = node.id,
        displayValue = node.props.getString("display_value"),
        label = node.props.getString("label"),
        helper = node.props.getString("helper"),
        error = node.props.getString("error"),
        disabled = node.props.getBool("disabled"),
        canDecrement = node.props.getBool("can_decrement"),
        canIncrement = node.props.getBool("can_increment"),
        decrementCallback = node.props.getCallbackId("on_decrement"),
        incrementCallback = node.props.getCallbackId("on_increment"),
        accessibilityLabel = node.props.getString("a11y_label").ifEmpty { node.props.getString("label") },
        accessibilityHint = node.props.getString("a11y_hint"),
    )

    val canPressDecrement: Boolean get() = !disabled && canDecrement && decrementCallback != 0
    val canPressIncrement: Boolean get() = !disabled && canIncrement && incrementCallback != 0
}

class StepperRendererState(configuration: StepperRendererConfiguration) {
    var configuration by mutableStateOf(configuration); private set
    var isAwaitingPublication by mutableStateOf(false); private set

    constructor(node: NativeUINode) : this(StepperRendererConfiguration(node))

    fun decrement(): StepperRendererEvent? = propose(
        callbackId = configuration.decrementCallback,
        enabled = configuration.canPressDecrement,
    )

    fun increment(): StepperRendererEvent? = propose(
        callbackId = configuration.incrementCallback,
        enabled = configuration.canPressIncrement,
    )

    fun serverPublished(tree: NativeUITree): Boolean {
        val node = tree.root.findStepperNode(configuration.nodeId) ?: return false
        val previousDisplayValue = configuration.displayValue
        configuration = StepperRendererConfiguration(node)
        isAwaitingPublication = false
        return previousDisplayValue != configuration.displayValue
    }

    fun serverPublished(published: StepperRendererConfiguration) {
        configuration = published
        isAwaitingPublication = false
    }

    private fun propose(callbackId: Int, enabled: Boolean): StepperRendererEvent? {
        if (isAwaitingPublication || !enabled) return null
        isAwaitingPublication = true
        return StepperRendererEvent(callbackId = callbackId, nodeId = configuration.nodeId)
    }
}

object StepperRenderer {
    @Composable
    fun Render(
        node: NativeUINode,
        modifier: Modifier,
        events: StepperRendererEvents = StepperRendererEvents.native,
    ) {
        val state = remember(node.id) { StepperRendererState(node) }

        LaunchedEffect(node.props, node.onPress) {
            state.serverPublished(StepperRendererConfiguration(node))
        }
        NativeUIBridge.currentTree.value?.let { tree ->
            LaunchedEffect(tree) { state.serverPublished(tree) }
        }

        FirstlightStepperControl(
            state = state,
            tokens = if (isSystemInDarkTheme()) NativeUITheme.dark else NativeUITheme.light,
            onDecrement = { state.decrement()?.let(events::dispatch) },
            onIncrement = { state.increment()?.let(events::dispatch) },
            modifier = modifier,
        )
    }
}

private fun NativeUINode.findStepperNode(id: Int): NativeUINode? {
    if (this.id == id) return this
    children.forEach { child -> child.findStepperNode(id)?.let { return it } }
    return null
}

File: resources/ios/StepperRenderer.swift
import SwiftUI

enum StepperRendererEvent: Equatable {
    case press(callbackId: Int, nodeId: Int)

    var wireName: String { "PRESS" }
}

struct StepperRendererEvents {
    let send: @MainActor (StepperRendererEvent) -> Void

    static let native = StepperRendererEvents { event in
        switch event {
        case let .press(callbackId, nodeId):
            NativeUIBridge.sendPressEvent(callbackId, nodeId: nodeId)
        }
    }
}

struct StepperRendererConfiguration: Equatable {
    let nodeID: Int
    let displayValue: String
    let label: String
    let helper: String
    let error: String
    let disabled: Bool
    let canDecrement: Bool
    let canIncrement: Bool
    let decrementCallback: Int
    let incrementCallback: Int
    let accessibilityLabel: String
    let accessibilityHint: String

    init(node: NativeUINode) {
        let props = node.props
        nodeID = node.id
        displayValue = props.getString("display_value")
        label = props.getString("label")
        helper = props.getString("helper")
        error = props.getString("error")
        disabled = props.getBool("disabled")
        canDecrement = props.getBool("can_decrement")
        canIncrement = props.getBool("can_increment")
        decrementCallback = props.getCallbackId("on_decrement")
        incrementCallback = props.getCallbackId("on_increment")
        let explicitLabel = props.getString("a11y_label")
        accessibilityLabel = explicitLabel.isEmpty ? label : explicitLabel
        accessibilityHint = props.getString("a11y_hint")
    }

    init(
        nodeID: Int,
        displayValue: String,
        label: String,
        helper: String,
        error: String,
        disabled: Bool,
        canDecrement: Bool,
        canIncrement: Bool,
        decrementCallback: Int,
        incrementCallback: Int,
        accessibilityLabel: String,
        accessibilityHint: String
    ) {
        self.nodeID = nodeID
        self.displayValue = displayValue
        self.label = label
        self.helper = helper
        self.error = error
        self.disabled = disabled
        self.canDecrement = canDecrement
        self.canIncrement = canIncrement
        self.decrementCallback = decrementCallback
        self.incrementCallback = incrementCallback
        self.accessibilityLabel = accessibilityLabel
        self.accessibilityHint = accessibilityHint
    }

    var canPressDecrement: Bool { !disabled && canDecrement && decrementCallback != 0 }
    var canPressIncrement: Bool { !disabled && canIncrement && incrementCallback != 0 }
}

struct StepperRendererState: Equatable {
    var configuration: StepperRendererConfiguration
    private(set) var isAwaitingPublication = false

    init(node: NativeUINode) {
        configuration = StepperRendererConfiguration(node: node)
    }

    init(configuration: StepperRendererConfiguration) {
        self.configuration = configuration
    }

    mutating func decrement() -> StepperRendererEvent? {
        propose(callbackId: configuration.decrementCallback, enabled: configuration.canPressDecrement)
    }

    mutating func increment() -> StepperRendererEvent? {
        propose(callbackId: configuration.incrementCallback, enabled: configuration.canPressIncrement)
    }

    @discardableResult
    mutating func serverPublished(_ tree: NativeUITree) -> Bool {
        guard let node = Self.findNode(id: configuration.nodeID, in: tree.root) else { return false }

        let previousDisplayValue = configuration.displayValue
        configuration = StepperRendererConfiguration(node: node)
        isAwaitingPublication = false

        return previousDisplayValue != configuration.displayValue
    }

    mutating func serverPublished(_ published: StepperRendererConfiguration) {
        configuration = published
        isAwaitingPublication = false
    }

    private mutating func propose(callbackId: Int, enabled: Bool) -> StepperRendererEvent? {
        guard !isAwaitingPublication, enabled else { return nil }
        isAwaitingPublication = true

        return .press(callbackId: callbackId, nodeId: configuration.nodeID)
    }

    private static func findNode(id: Int, in node: NativeUINode) -> NativeUINode? {
        if node.id == id { return node }
        for child in node.children {
            if let found = findNode(id: id, in: child) { return found }
        }
        return nil
    }
}

struct StepperRenderer: View {
    @ObservedObject private var themeStore = NativeUITheme.shared
    @ObservedObject private var bridge = NativeUIBridge.shared
    @Environment(\.colorScheme) private var colorScheme
    @State private var state: StepperRendererState
    private let events: StepperRendererEvents

    init(node: NativeUINode, events: StepperRendererEvents = .native) {
        _state = State(initialValue: StepperRendererState(node: node))
        self.events = events
    }

    var body: some View {
        FirstlightStepperControl(
            state: $state,
            tokens: themeStore.resolve(for: colorScheme),
            onEvent: events.send
        )
        .onReceive(bridge.$currentTree.compactMap { $0 }) { tree in
            state.serverPublished(tree)
        }
    }
}
